File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <iostream>
using namespace std;

int answer = 0, t, n;
vector<int> arr;

void dfs(int cnt, int sum){
    if(cnt == n){
        if(sum == t) answer++;
        return;
    }
    
    dfs(cnt + 1, sum + arr[cnt]);
    dfs(cnt + 1, sum - arr[cnt]);
}

int solution(vector<int> numbers, int target) {
    arr = numbers;
    t = target;
    n = numbers.size();
    
    dfs(0, 0);
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
import java.util.*;

class Solution {
    public int answer = 0;
    
    public int solution(int[] numbers, int target) {
        dfs(numbers, target, 0, 0);
        
        return answer;
    }
    
    public void dfs(int[] numbers, int target, int sum, int cnt) {
        if(cnt == numbers.length) {
            if(sum == target) answer++;
            return;
        }
        dfs(numbers, target, sum + numbers[cnt], cnt + 1);
        dfs(numbers, target, sum - numbers[cnt], cnt + 1);
    }
}